File: golang/add-two-numbers.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var head *ListNode
	var tail *ListNode
	carry := false
	for l1 != nil || l2 != nil {
		num := 0
		if l1 != nil && l2 != nil {
			num = l1.Val + l2.Val
			l1 = l1.Next
			l2 = l2.Next
		} else if l1 != nil {
			num = l1.Val
			l1 = l1.Next
		} else {
			num = l2.Val
			l2 = l2.Next
		}
		if carry {
			num++
			carry = false
		}
		if num > 9 {
			num %= 10
			carry = true
		}
		if head == nil {
			head = &ListNode{num, nil}
			tail = head
		} else {
			tail.Next = &ListNode{num, nil}
			tail = tail.Next
		}
	}
	if carry {
		tail.Next = &ListNode{1, nil}
	}
	return head
}

File: python/add-two-numbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        dummy=tail=ListNode(0)
        carry=False
        while l1 is not None or l2 is not None:
            num=0
            if l1 is not None:
                num+=l1.val
                l1=l1.next
            if l2 is not None:
                num+=l2.val
                l2=l2.next
            if carry:
                num+=1
                carry=False
            if num>9:
                num%=10
                carry=True
            tail.next=ListNode(num)
            tail=tail.next
        if carry:
            tail.next=ListNode(1)
        return dummy.next
